### PR TITLE
caml_output_value_to_bytes: avoid integer overflow

### LINKED ISSUE #8787: Segmentation fault: Marshal.to_string with large float32 bigarray
Minimal reproduction steps:
```
(* in the toplevel *)
open Bigarray;;
let m = Genarray.create float32 c_layout [|58320; 11664|];;
Marshal.to_string m [];;
```

Note that it does fail also if `float32` is replaced by `complex64`, but does not fail with `float64` and `complex32`. Tested on mac osx 10.14.5, with ocaml 4.07.1 (also in native mode).

Origin: https://github.com/owlbarn/owl/issues/422

### PR DESCRIPTION
**Pull request by yallop**
Use `intnat`, not `int`, to represent pointer difference.

Fixes #8787

### CODE REVIEW

**Review by dra27 — APPROVED** [review]
Just needs a changes entry, please.

**Comment by yallop** [comment]
There's a `Changes` entry in 2aa1b3f .

**Comment by yallop** [comment]
I've added another commit for what appears to be the same problem in `caml_output_value_to_malloc` (fa9051a).  I haven't tested the change, because that function doesn't appear to be used internally.

**Comment by yallop** [comment]
@dra27: do you still approve after the latest change?  If so, this can be merged.

**Comment by dra27** [comment]
Yes, I do! @damiendoligez - 4.09 and 4.08? (I'll move the `Changes` entry to the right place before merging)

**Comment by mseri** [comment]
Thanks for the fast fix!

**Comment by damiendoligez** [comment]
@dra27 cherry-pick to 4.09, yes. Don't bother with 4.08, thanks.

**Pull request merged**
